### PR TITLE
Enable explicit use of key tuples (instead of *Indexer objects) in indexing adapters and explicitly indexed arrays

### PR DESCRIPTION
**Pull request by andersy005**
- [ ] Towards #8856
- [ ] Tests added
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] New functions/methods are listed in `api.rst`

### CODE REVIEW

File: xarray/coding/strings.py [hunk 1]
```diff
@@ -249,15 +250,15 @@ def shape(self) -> tuple[int, ...]:
     def __repr__(self):
         return f"{type(self).__name__}({self.array!r})"
 
-    def _vindex_get(self, key):
+    def _vindex_get(self, key: tuple[Any, ...]):
```
Comment (Illviljan) [inline xarray/coding/strings.py:253]: A lot of `tuple[Any, ...]`! What do you think of using variable name instead, `_IndexerKey = tuple[Any, ...]`? The intention is they should all have the same typing, right?

I worry when it starts repeating this much there's a risk the types starts diverging in future PRs.

File: xarray/coding/strings.py [hunk 2]
```diff
@@ -221,7 +222,7 @@ class StackedBytesArray(indexing.ExplicitlyIndexedNDArrayMixin):
     values, when accessed, are automatically stacked along the last dimension.
 
     >>> indexer = indexing.BasicIndexer((slice(None),))
-    >>> StackedBytesArray(np.array(["a", "b", "c"], dtype="S1"))[indexer]
+    >>> StackedBytesArray(np.array(["a", "b", "c"], dtype="S1"))[indexer.tuple]
```
Comment (dcherian) [inline xarray/coding/strings.py:225]:
```suggestion
    >>> StackedBytesArray(np.array(["a", "b", "c"], dtype="S1"))[slice(None)]
```

let's also delete the previous line. These docstrings won't have access to `*Indexer`

File: xarray/coding/variables.py [hunk 3]
```diff
@@ -100,7 +100,7 @@ class NativeEndiannessArray(indexing.ExplicitlyIndexedNDArrayMixin):
     dtype('int16')
 
     >>> indexer = indexing.BasicIndexer((slice(None),))
-    >>> NativeEndiannessArray(x)[indexer].dtype
+    >>> NativeEndiannessArray(x)[indexer.tuple].dtype
```
Comment (dcherian) [inline xarray/coding/variables.py:103]: Same here.

File: xarray/core/indexing.py [hunk 4]
```diff
@@ -517,29 +517,29 @@ class ExplicitlyIndexedNDArrayMixin(NDArrayMixin, ExplicitlyIndexed):
 
     def get_duck_array(self):
         key = BasicIndexer((slice(None),) * self.ndim)
-        return self[key]
+        return self[key.tuple]
```
Comment (dcherian) [inline xarray/core/indexing.py:520]: Again here.

File: xarray/tests/test_backends.py [hunk 5]
```diff
@@ -5811,7 +5811,7 @@ def _getitem(self, key):
 
     with warnings.catch_warnings(record=True) as w:
         warnings.simplefilter("always")
-        la.vindex[indexer].get_duck_array()
+        la.vindex[indexer.tuple].get_duck_array()
```
Comment (dcherian) [inline xarray/tests/test_backends.py:5814]:
```suggestion
        la[indexer].get_duck_array()
```

this seems like a good place to check for backcompat warnings in custom backends.

File: xarray/tests/test_coding_strings.py [hunk 6]
```diff
@@ -181,7 +181,7 @@ def test_StackedBytesArray_vectorized_indexing() -> None:
 
     V = IndexerMaker(indexing.VectorizedIndexer)
     indexer = V[np.array([[0, 1], [1, 0]])]
-    actual = stacked.vindex[indexer]
+    actual = stacked.vindex[indexer.tuple]
```
Comment (dcherian) [inline xarray/tests/test_coding_strings.py:184]: Just pass the tuple directly?

File: xarray/tests/test_dataset.py [hunk 7]
```diff
@@ -5070,28 +5070,26 @@ def test_lazy_load(self) -> None:
...
-            assert isinstance(ds["var1"].load().data, DuckArrayWrapper)
-            assert isinstance(
-                ds["var1"].isel(dim2=0, dim1=0).load().data, DuckArrayWrapper
-            )
+        # preserve the duck array type and don't cast to array
+        assert isinstance(ds["var1"].load().data, DuckArrayWrapper)
+        assert isinstance(ds["var1"].isel(dim2=0, dim1=0).load().data, DuckArrayWrapper)
 
-            ds.close()
+        ds.close()
```
Comment (dcherian) [inline xarray/tests/test_dataset.py:5092]: let's indent these back. I think they're testing the indexing with a number of "decoding" array types

File: xarray/core/indexing.py [hunk 8]
```diff
@@ -1040,29 +1037,63 @@ def explicit_indexing_adapter(
     return result
 
 
+class CompatIndexedTuple(tuple):
```
Comment (dcherian) [inline xarray/core/indexing.py:1040]: Nice!

File: xarray/core/indexing.py [hunk 9]
```diff
@@ -1728,13 +1752,14 @@ def _convert_scalar(self, item):
         # a NumPy array.
         return to_0d_array(item)
 
-    def _prepare_key(self, key: tuple[Any, ...]) -> tuple[Any, ...]:
-        if isinstance(key, tuple) and len(key) == 1:
+    def _prepare_key(self, key: _IndexerKey) -> _IndexerKey:
```
Comment (dcherian) [inline xarray/core/indexing.py:1755]:
```suggestion
    def _prepare_key(self, key: ExplicitIndexer | _IndexerKey) -> _IndexerKey:
```

File: xarray/tests/test_dataset.py (hunk 7 above)
Comment (andersy005) [reply xarray/tests/test_dataset.py:5092]: i had parametrized them -- which is why i removed the loop/indentation

Comment (dcherian) [reply xarray/tests/test_dataset.py:5092]: ah nice!

File: xarray/core/indexing.py [hunk 10]
```diff
@@ -1040,29 +1036,63 @@ def explicit_indexing_adapter(
...
+    backend arrays that currently expect this metadata to function correctly while
+    maintaining the outward behavior of a regular tuple.
+
+    This class is particularly useful during the phase where the backend implementations are
+    not yet capable of directly accepting raw tuples without additional context about
+    the indexing type. It ensures that these backends can still correctly interpret and
+    process indexing operations by providing them with the necessary contextual information.
+    """
+
+    def __new__(cls, iterable, indexer_type: Literal["basic", "vectorized", "outer"]):
```
Comment (dcherian) [inline xarray/core/indexing.py:1053]: Let's make `indexer_type` an `class IndexerType(Enum)`

File: xarray/coding/strings.py [hunk 11]
```diff
@@ -249,15 +249,17 @@ def shape(self) -> tuple[int, ...]:
...
+        from xarray.core.indexing import CompatIndexedTuple
+
         # require slicing the last dimension completely
-        key = type(key)(indexing.expanded_indexer(key.tuple, self.array.ndim))
-        if key.tuple[-1] != slice(None):
+        indexer = indexing.expanded_indexer(key, self.array.ndim)
+        if indexer[-1] != slice(None):
             raise IndexError("too many indices")
-        return _numpy_char_to_bytes(self.array[key])
+        return _numpy_char_to_bytes(self.array[CompatIndexedTuple(indexer, "basic")])
```
Comment (dcherian) [inline xarray/coding/strings.py:265]:
```suggestion
        return _numpy_char_to_bytes(self.array[BasicIndexer(indexer)])
```

File: xarray/coding/variables.py [hunk 12]
```diff
@@ -138,7 +137,7 @@ class BoolTypeArray(indexing.ExplicitlyIndexedNDArrayMixin):
     dtype('bool')
 
     >>> indexer = indexing.BasicIndexer((slice(None),))
-    >>> BoolTypeArray(x)[indexer].dtype
+    >>> BoolTypeArray(x)[indexer.tuple].dtype
```
Comment (dcherian) [inline xarray/coding/variables.py:140]: another docstring that can be updated

File: xarray/tests/test_backends.py (hunk 5 above)
Comment (andersy005) [reply xarray/tests/test_backends.py:5814]: i intend to look into this later today

File: xarray/core/indexing.py [hunk 13]
```diff
@@ -673,31 +672,32 @@ def get_duck_array(self):
...
         return array.vindex[indexer]
 
-    def __getitem__(self, indexer: ExplicitIndexer):
-        self._check_and_raise_if_non_basic_indexer(indexer)
-        return type(self)(self.array, self._updated_key(indexer))
+    def __getitem__(self, indexer: _IndexerKey):
+        key = BasicIndexer(
+            indexer.tuple if isinstance(indexer, ExplicitIndexer) else indexer
+        )
+        return type(self)(self.array, self._updated_key(key))
```
[the remark this answers is not in the document]
Comment (andersy005) [reply xarray/core/indexing.py]: added some initial updates with support for raw tuples as input. however, there's still use of `indexer` objects within the method itself. my attempts at completely removing indexer objects resulted in plenty of issues that i wasn't ready to address :). will look into this later today as well

**Comment by dcherian** [comment]
Very nice work, Anderson!

We can address the backend warnings in the next PR

**Pull request merged**